yabt/dot.py

```diff
@@ -0,0 +1,79 @@
+# -*- coding: utf-8 -*-
+
+# Copyright 2019 Resonai Ltd. All rights reserved
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Creates the dot graph
+~~~~~~~~~~~~~~~~~
+
+:author: Dana Shamir
+"""
+
+
+from .caching import get_prebuilt_targets
+from .config import Config
+
+
+TARGETS_COLORS = {'AptPackage': 'brown4',
+                  'CppGTest': 'deepskyblue4',
+                  'CppLib': 'blue',
+                  'CppProg': 'cornflowerblue',
+                  'CustomInstaller': 'brown',
+                  'Proto': 'green',
+                  'Python': 'red',
+                  'PythonPackage': 'purple',
+                  'PythonTest': 'pink'
+                  }
+
+
+def get_not_buildenv_targets(build_context):
+    roots = [n for n, d in build_context.target_graph.in_degree() if d == 0]
+    buildenvs = set(
+        target.buildenv for target in build_context.targets.values()
+        if target.buildenv is not None)
+    visited = set()
+    to_do = []
+    for root in roots:
+        if root not in buildenvs:
+            to_do.append(root)
+
+    while to_do:
+        node = to_do.pop(-1)
+        if node not in visited:
```

Comment:
FYI. that's the ugly (performance-wise) way to do it: you are checking if something is visited after popping instead of before pushing. for large graphs this is really bad. Here, I'm not sure, but it's bad form.